### DISCUSSION TITLE
Preconfigured command configuration and rules in binary?

### DISCUSSION THREAD

**Discussion by bendecaf** [arendst/Tasmota#11408] [Q&A] [answered]
Hi,

I'm trying to find a way to preconfigure a tasmota binary for my little sensor device that I have built. I have found most of what I need in my_user_config.h and override it in user_config_override.h, i.e., I have configured my user template, enable the display and sensor that I need etc. What I haven't figured out yet is how I can preconfigure DisplayAddress, DisplayModel, DisplayMode and the rules in the binary that I currently send out through commands in the console. Is it possible to preconfigure those as well and if so, how does that work?

Thanks for your help!

**Comment by sfromis** [comment]
There is no feature for including commands or rules in the precompiled binary.

**Reply by bendecaf** [reply]
I suspect this is unfortuantely the most correct answer with respect to the intent of my question :-(

**Comment by barbudor** [comment]
Still some settings can be hardocded in the binary.
Check `my_user_config.h` for a non exhaustive list.
You can copy/paste settings from there into your `user_config_override.h` to change, re-define, ...
If a setting you are looking for is missing, let us know. It could be possible/considered to add it to the list.

**Comment by barbudor** [comment]
Alternatively you can save the configuration from an existing device as a JSON file using `decode-config`
Then strip down the JSON to the settings that you need (for example the rules)
You can then use that JSON with `decode-config` to send back to other devices

**Reply by bendecaf** [reply]
Thanks for the pointer. I guess it is useful, but if I understand correctly, it doesn't help me to preconfigure the firmware with this reduced json config. I kind of want to preconfigure tasmota for usage of a very specific pcb so that possible users only have to to care of the wifi and mqtt setup. Everything else should be preconfigured. If there was one (or multiple) simple #define statements in the user_config_override.h where I could add such a reduced json set or a backlog statement which is setup on initial startup, that would be ideal or kind of the the thing that I'm looking for, but I understand it is currently not possible.

**Comment by Jason2866** [comment, marked as answer]
Do you have 4MB flash? If yes uncomment this line https://github.com/arendst/Tasmota/blob/development/platformio_override_sample.ini#L61 
Now you have room for Tasmota file system and you can enable it via 
```
#define USE_UFILESYS
  #define GUI_TRASH_FILE
```
In your `user_config_override.h`
Place the file `autoexc.bat` in filesystem in which your commands are stored.
Every line one command. You can use every command which you can use in the console.
Only commands which are forcing a restart are NOT possible.

**Reply by bendecaf** [reply]
This looks very interesting. I almost thought that this is the solution, but it seems it's not possible to prepackage the autoexec.bat into the firmware, correct?

**Reply by barbudor** [reply]
Have a look at that : https://docs.platformio.org/en/latest/platforms/espressif8266.html#uploading-files-to-filesystem
You should be able to build an image of a filesystem with your file
Then you would be able to flash simultaneously the firmware and the filesystem
Never tried it myself

Beware that by default the autoexec.bat will execute again and again at each restart.
Which means it will re-execute the commands everytime, extending the boot time.

You may also want to start the autoexec.bat with a `savedata 0` to avoid writing the flash at each step of the configuration, and erstore `savedata 1` at the end

**Reply by bendecaf** [reply]
Ok thanks. I'll try that. Thanks a lot for your help!

**Reply by Jason2866** [reply]
I have implemented the file upload function in VSC/Platformio already.
Just make a folder `data` in the root of the project and place your `autoxec.bat` there.
See https://github.com/arendst/Tasmota/pull/10533
